### DISCUSSION TITLE
Maybe bug: Style classes in global scope

### DISCUSSION THREAD

**Discussion by buhtz** [pygments/pygments#2390] [Q&A] [answered]
By default pygments code blocks are wrapped in a css class named "highlight"

```
<div class="highlight"><pre>code
</pre></div>
```

But the resulting css file do not use that class.

```
pre { line-height: 125%; }
td.linenos .normal { color: inherit; background-color: transparent; padding-left: 5px; padding-right: 5px; }
span.linenos { color: inherit; background-color: transparent; padding-left: 5px; padding-right: 5px; }
td.linenos .special { color: #000000; background-color: #ffffc0; padding-left: 5px; padding-right: 5px; }
span.linenos.special { color: #000000; background-color: #ffffc0; padding-left: 5px; padding-right: 5px; }
body .hll { background-color: #ffffcc }
body { background: #f8f8f8; }
body .c { color: #3D7B7B; font-style: italic } /* Comment */
body .err { border: 1px solid #FF0000 } /* Error */
body .k { color: #008000; font-weight: bold } /* Keyword */
```

The classes (.c, .k, ...) are valid everywhere in the <body> and not only in <div class="highlight">.

This can cause problems.

It is not hard to workaround that and fix the CSS myself but it could be a new feature to make it possible to restrict the CSS to the scope of a specific class.

**Comment by birkenfeld** [comment, marked as answer]
You can pass the desired CSS prefix(es) as an argument to `HtmlFormatter.get_style_defs`. I'd be wary of changing the default due to compatibility concerns.

**Reply by buhtz** [reply]
Ah! I was looking in the "docu" of HtmlFormater.__init__() instead of that method. My fault.
Glad asked before openinga bug report. 😆

**Reply by buhtz** [reply]
The docu doesn't help much.

https://pygments.org/docs/api/#pygments.formatter.Formatter.get_style_defs

It tells me the arguments are formatter depended. OK, but I didn't found that in the formatter.

You really have to overthink your documentation to prevent users like me asking such stupid question that could be answered with a usual API docu.

**Reply by jeanas** [reply]
> OK, but I didn't found that in the formatter.

On https://pygments.org/docs/formatters/#HtmlFormatter it says

> The get_style_defs(arg=’’) method of a HtmlFormatter returns a string containing CSS rules for the CSS classes used by the formatter. The argument arg can be used to specify additional CSS selectors that are prepended to the classes. A call fmter.get_style_defs(‘td .code’) would result in the following CSS classes:
> 
> td .code .kw { font-weight: bold; color: #00FF00 }
> td .code .cm { color: #999999 }
> ...
>
> If you have Pygments 0.6 or higher, you can also pass a list or tuple to the get_style_defs() method to request multiple prefixes for the tokens:
>
> formatter.get_style_defs(['div.syntax pre', 'pre.syntax'])
>
> The output would then look like this:
>
> div.syntax pre .kw,
> pre.syntax .kw { font-weight: bold; color: #00FF00 }
> div.syntax pre .cm,
> pre.syntax .cm { color: #999999 }
> ...

**Reply by buhtz** [reply]
1. You can not put the docu for a method into the class docu.
2. The string "additional CSS selectors that are prepended to the classes" doesn't explain what really happens. While reading this I thought it just select a subset from the existing CSS and not adding something new (a "prefix") to the CSS. The examples in there doesn't help if the users doesn't know how the default CSS do look like.

**Reply by birkenfeld** [reply]
Suggestions for better wording are welcome.
